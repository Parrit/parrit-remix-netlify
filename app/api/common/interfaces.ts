export interface LoginRequest {
  projectName?: string;
  password?: string;
}

export interface LoginResult {
  message: string;
}

export interface Project {
  id: string;
  name: string;
  pairingBoards: PairingBoard[];
  people: Person[];
}

export interface PairingBoard {
  id: string;
  name: string;
  exempt: boolean;
  people: Person[];
  roles: Role[];
}

export interface Person {
  id: string;
  name: string;
}

export interface Role {
  id: string;
  name: string;
}

export interface PairingArrangementDTO {
  id: number; // this should be changed to string
  pairingTime: string;
  pairingHistories: PairingHistory[];
}

export interface PairingHistory {
  pairingBoardName: string;
  people: Person[];
}
