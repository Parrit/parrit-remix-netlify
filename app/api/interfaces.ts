export interface FormErrors<T> {
  server?: string;
  fields?: Record<keyof T, string | null>;
  status: 400 | 401 | 500;
}

export const isFormErrors = <T>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  arg: FormErrors<T> | any
): arg is FormErrors<T> => {
  return (arg as FormErrors<T>).status > 0;
};
