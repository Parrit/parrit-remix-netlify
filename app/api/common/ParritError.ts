export interface IParritError<T> {
  server?: string;
  fields?: Record<keyof T, string | null>;
}

export class ParritError<T> extends Error {
  public readonly data: IParritError<T>;

  constructor(arg: IParritError<T>) {
    super(JSON.stringify(arg));
    this.data = arg;
  }

  toResponse(status = 400): Response {
    return new Response(JSON.stringify(this.data), { status });
  }

  static obscured() {
    return new ParritError({ server: "Server error" });
  }

  static fromString<T>(str?: string) {
    if (!str) {
      return undefined;
    }
    const obj = JSON.parse(str) as IParritError<T>;
    if (this.isParritError<T>) {
      return new ParritError(obj);
    }
    return undefined;
  }

  private static isParritError<T>(obj: object): obj is ParritError<T> {
    const cast = obj as ParritError<T>;
    return !!cast.data.server || !!cast.data.fields;
  }
}
