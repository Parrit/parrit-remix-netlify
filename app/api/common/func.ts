import {
  FLOATING_IDX,
  PairingBoard,
  Person,
  Project,
} from "./interfaces/parrit.interfaces";
import * as R from "ramda";

// Functional programing utilities for working with interfaces

export const movePerson = (
  project: Project,
  person: Person,
  position: PairingBoard
) =>
  R.pipe<[Project], Project, Project>(
    removePerson(person),
    addPerson(person, position)
  )(project);

export const addPerson = R.curry(
  (person: Person, position: PairingBoard, project: Project) => {
    const f1 = addPersonToPosition(person, position);
    const f2 =
      position.id === FLOATING_IDX ? addPersonToFloating(person) : R.identity;
    return R.pipe<[Project], Project, Project>(f1, f2)(project);
  }
);

const addPersonToPosition = R.curry(
  (person: Person, position: PairingBoard, project: Project) =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      R.map<PairingBoard, PairingBoard>((x) => {
        if (x.id === position.id) {
          return R.over<PairingBoard, Person[]>(
            R.lensProp("people"),
            R.append({ ...person, pairing_board_id: position.id }),
            x
          );
        }
        return x;
      }),
      project
    )
);

const addPersonToFloating = R.curry((person: Person, project: Project) =>
  R.over<Project, Person[]>(
    R.lensPath(["floating", "people"]),
    R.append(person),
    project
  )
);

export const removePerson = R.curry(
  (person: Person, project: Project): Project =>
    R.pipe<[Project], Project, Project>(
      removePersonFromBoards(person),
      removePersonFromFloating(person)
    )(project)
);

const removePersonFromBoards = R.curry(
  (person: Person, project: Project): Project =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      R.map<PairingBoard, PairingBoard>((x) => {
        if (x.people.includes(person)) {
          return R.set<PairingBoard, Person[]>(
            R.lensProp("people"),
            R.without(x.people, [person]),
            x
          );
        }
        return x;
      }),
      project
    )
);

const removePersonFromFloating = R.curry(
  (person: Person, project: Project): Project =>
    R.set<Project, Person[]>(
      R.lensPath(["floating", "people"]),
      R.without([person], project.floating.people),
      project
    )
);

const pbIndexById = R.curry((pbs: PairingBoard[], pb_id: string) =>
  R.findIndex<PairingBoard>((pb) => pb.id === pb_id)(pbs)
);

const renamePairingBoardFromList = R.curry(
  (pb_id: string, rename: string, pbs: PairingBoard[]) => {
    const foundIndex = pbIndexById(pbs, pb_id);
    if (foundIndex < 0) {
      return pbs;
    }
    return R.adjust<PairingBoard>(foundIndex, R.assoc("name", rename), pbs);
  }
);

export const renamePairingBoard = R.curry(
  (project: Project, pb_id: string, rename: string): Project =>
    R.over<Project, PairingBoard[]>(
      R.lensProp("pairingBoards"),
      renamePairingBoardFromList(pb_id, rename),
      project
    )
);

export const canAPairingBeMade = (project: Project): boolean => {
  const atLeast2Floaters = project.floating.people.length >= 2;
  const atLeast1Floater = project.floating.people.length >= 1;
  const atLeast1EmptyBoard = getEmptyPairingBoard(project) !== undefined;
  const atLeast1UnpairedSticker = unpairedStickingPeople(project).length >= 1;
  return (
    (atLeast1Floater && atLeast1UnpairedSticker) ||
    (atLeast2Floaters && atLeast1EmptyBoard)
  );
};

export const getEmptyPairingBoard = (
  project: Project
): PairingBoard | undefined => {
  return project.pairingBoards.find(
    (pb) => pb.people.length === 0 && !pb.exempt
  );
};

const unpairedStickingPeople = (project: Project): Person[] =>
  project.pairingBoards
    .filter((board) => !board.exempt)
    .filter((board) => board.people.length === 1)
    .flatMap((board) => board.people);

const findPairingBoardByPerson = (
  project: Project,
  person: Person
): PairingBoard | undefined =>
  project.pairingBoards.find((pb) => pb.people.find((p) => p.id === person.id));
