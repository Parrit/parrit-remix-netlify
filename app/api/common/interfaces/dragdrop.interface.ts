export type DragType = "Person" | "Role";

export enum DropType {
  Floating = "Floating",
  PairingBoard = "PairingBoard",
  TrashBin = "TrashBin",
}

export interface DragItem {
  id: string;
  name: string;
  type: DragType;
  pairing_board_id: string;
}

export interface DropItem {
  type: DropType;
}

export interface DropTarget {
  isOver: boolean;
  canDrop: boolean;
}

export interface DragResult {}
