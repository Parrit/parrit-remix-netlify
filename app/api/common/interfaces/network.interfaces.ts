export interface LoginRequest {
  projectName?: string;
  password?: string;
}

export interface LoginResult {
  message: string;
}
