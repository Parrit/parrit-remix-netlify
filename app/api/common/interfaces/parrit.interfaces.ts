import { DragItem } from "~/api/common/interfaces/dragdrop.interface";

export interface Project {
  id: string;
  name: string;
  pairingBoards: PairingBoard[];
  floating: PairingBoard;
  people: Person[]; // the new cannonical location for people
  roles: Role[]; // the new cannonical location for roles
}

export const FLOATING_IDX = "-1";

export interface PairingBoard {
  id: string;
  name: string;
  exempt: boolean;
}

export interface Person extends DragItem {
  id: string;
  name: string;
  type: "Person";
}

export interface Role extends DragItem {
  id: string;
  name: string;
  type: "Role";
}

export interface ProjectPairingSnapshot {
  id: string;
  projectId: string;
  pairingTime: string;
  pairingInstances: PairingInstance[];
}

export interface PairingInstance {
  projectId: string;
  pairingBoardName: string | null;
  people: Person[];
  pairingTime: string;
}
