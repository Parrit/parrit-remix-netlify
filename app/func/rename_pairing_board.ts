import { curry, over, lensProp, findIndex, adjust, assoc } from "ramda";
import {
  Project,
  PairingBoard,
} from "~/api/common/interfaces/parrit.interfaces";

export default curry(
  (project: Project, pb_id: string, rename: string): Project =>
    over<Project, PairingBoard[]>(
      lensProp("pairingBoards"),
      rename_pairing_board_from_list(pb_id, rename),
      project
    )
);

const pb_index_by_id = curry((pbs: PairingBoard[], pb_id: string) =>
  findIndex<PairingBoard>((pb) => pb.id === pb_id)(pbs)
);

const rename_pairing_board_from_list = curry(
  (pb_id: string, rename: string, pbs: PairingBoard[]) => {
    const foundIndex = pb_index_by_id(pbs, pb_id);
    if (foundIndex < 0) {
      return pbs;
    }
    return adjust<PairingBoard>(foundIndex, assoc("name", rename), pbs);
  }
);
