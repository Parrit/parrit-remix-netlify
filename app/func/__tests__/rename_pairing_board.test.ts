import { expect } from "@jest/globals";
import { FLOATING_IDX } from "~/api/common/interfaces/parrit.interfaces";
import { PROJECT_START, PB_RED } from "./common";
import rename_pairing_board from "../rename_pairing_board";

describe("rename_paring_board", () => {
  it("happy path", () => {
    const result = rename_pairing_board(PROJECT_START, PB_RED.id, "Frog Board");
    expect(result.pairingBoards[0].name).toBe("Frog Board");
    expect(result).toMatchSnapshot();
  });

  it("does not rename the floating board", () => {
    const result = rename_pairing_board(
      PROJECT_START,
      FLOATING_IDX,
      "NO, DONT"
    );
    expect(result).toMatchSnapshot();
  });
});
