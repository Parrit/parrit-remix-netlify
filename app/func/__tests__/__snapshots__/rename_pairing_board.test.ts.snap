// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rename_paring_board does not rename the floating board 1`] = `
{
  "floating": {
    "exempt": false,
    "id": "-1",
    "name": "floating_board_name",
  },
  "id": "project_id",
  "name": "project_name",
  "pairingBoards": [
    {
      "exempt": false,
      "id": "red_id",
      "name": "red_start",
    },
    {
      "exempt": false,
      "id": "blue_id",
      "name": "blue_start",
    },
    {
      "exempt": true,
      "id": "exempt",
      "name": "out_of_office",
    },
  ],
  "people": [
    {
      "id": "red_parrit_id",
      "name": "red_parrit",
      "pairing_board_id": "red_id",
      "type": "Person",
    },
    {
      "id": "99",
      "name": "Ms. Frizzle",
      "pairing_board_id": "brand new individual",
      "type": "Person",
    },
    {
      "id": "absent",
      "name": "Permanent Vacation",
      "pairing_board_id": "exempt",
      "type": "Person",
    },
    {
      "id": "floater_id",
      "name": "floater_name_start",
      "pairing_board_id": "-1",
      "type": "Person",
    },
  ],
  "roles": [
    {
      "id": "blue_role",
      "name": "Blue Role",
      "pairing_board_id": "blue_id",
      "type": "Role",
    },
    {
      "id": "ooo",
      "name": "OOOROLE",
      "pairing_board_id": "exempt",
      "type": "Role",
    },
  ],
}
`;

exports[`rename_paring_board happy path 1`] = `
{
  "floating": {
    "exempt": false,
    "id": "-1",
    "name": "floating_board_name",
  },
  "id": "project_id",
  "name": "project_name",
  "pairingBoards": [
    {
      "exempt": false,
      "id": "red_id",
      "name": "Frog Board",
    },
    {
      "exempt": false,
      "id": "blue_id",
      "name": "blue_start",
    },
    {
      "exempt": true,
      "id": "exempt",
      "name": "out_of_office",
    },
  ],
  "people": [
    {
      "id": "red_parrit_id",
      "name": "red_parrit",
      "pairing_board_id": "red_id",
      "type": "Person",
    },
    {
      "id": "99",
      "name": "Ms. Frizzle",
      "pairing_board_id": "brand new individual",
      "type": "Person",
    },
    {
      "id": "absent",
      "name": "Permanent Vacation",
      "pairing_board_id": "exempt",
      "type": "Person",
    },
    {
      "id": "floater_id",
      "name": "floater_name_start",
      "pairing_board_id": "-1",
      "type": "Person",
    },
  ],
  "roles": [
    {
      "id": "blue_role",
      "name": "Blue Role",
      "pairing_board_id": "blue_id",
      "type": "Role",
    },
    {
      "id": "ooo",
      "name": "OOOROLE",
      "pairing_board_id": "exempt",
      "type": "Role",
    },
  ],
}
`;
